components/authz-service/server/v2/rules_property_test.go

```diff
@@ -18,17 +21,511 @@ import (
+					ruleMatches(reqs.rules[0], *rStaged.Rule)
+			},
+			createProjectAndRuleGen,
+		))
+
+	properties.Property("creates a staged rule and applies it then confirms it is applied",
+		prop.ForAll(
+			func(reqs projectAndRuleReq) bool {
+				defer testDB.Flush(t)
+
+				_, err := cl.CreateProject(ctx, &reqs.CreateProjectReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				_, err = cl.CreateRule(ctx, &reqs.rules[0])
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				cl.ApplyRulesStart(ctx, &api.ApplyRulesStartReq{})
+
+				rApplied, err := cl.GetRule(ctx, &api.GetRuleReq{Id: reqs.rules[0].Id})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				return rApplied.Rule.Status == "applied" &&
+					ruleMatches(reqs.rules[0], *rApplied.Rule)
+			},
+			createProjectAndRuleGen,
+		))
+
+	properties.Property("ensure creating a rule fails if the rule ID is not unique",
+		prop.ForAll(
+			func(reqs projectAndRuleReq) bool {
+				defer testDB.Flush(t)
+
+				_, err := cl.CreateProject(ctx, &reqs.CreateProjectReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				if _, err := cl.CreateRule(ctx, &reqs.rules[0]); err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				_, err = cl.CreateRule(ctx, &reqs.rules[1])
+				return grpctest.AssertCode(t, codes.AlreadyExists, err)
+			},
+			// TODO(sr): deduplicate with the combinegen above?
+			gopter.CombineGens(
+				createProjectReqGen, createRuleReqGen, createRuleReqGen,
+			).Map(func(vals []interface{}) projectAndRuleReq {
+				p := vals[0].(api.CreateProjectReq)
+				// we "fix" the second req to use the same ID as the first one
+				r0 := vals[1].(api.CreateRuleReq)
+				r1 := vals[2].(api.CreateRuleReq)
+				r1.Id = r0.Id
+				r0.ProjectId = p.Id
+				r1.ProjectId = p.Id
+				return projectAndRuleReq{
+					CreateProjectReq: p,
+					rules:            []api.CreateRuleReq{r0, r1},
+				}
+			}),
+		))
+
+	properties.TestingRun(t)
+}
+
+func TestListRuleProperties(t *testing.T) {
+	ctx := context.Background()
+	cl, testDB, _, _, seed := testhelpers.SetupProjectsAndRulesWithDB(t)
+	properties := getGopterParams(seed)
+	_, _, createProjectAndRuleGen := getGenerators()
+
+	properties.Property("reports rules when all are staged and staged included in request",
+		prop.ForAll(
+			func(reqs projectAndRuleReq) bool {
+				defer testDB.Flush(t)
+
+				reportRules(t, reqs.rules)
+
+				_, err := cl.CreateProject(ctx, &reqs.CreateProjectReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				reqs.rules, err = createRules(ctx, cl, reqs.rules)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				resp, err := cl.ListRules(ctx, &api.ListRulesReq{IncludeStaged: true})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				return rulesMatch(reqs.rules, resp)
+
+			},
+			createProjectAndRuleGen,
```

Comment:
I guess we could have done something fancier here -- like, create a bunch of rules, apply some  of them, assert that numbers match.